Mage.Sets/src/mage/cards/g/GandalfWhiteRider.java

```diff
@@ -0,0 +1,135 @@
+        // When Gandalf, White Rider dies, you may put it into its owner's library fifth from the top.
+
+        // Adds ability that whenever you cast a spell it gives a temporary +1/+0 effect until end of turn
+        Ability ability = new SpellCastControllerTriggeredAbility(
+                new BoostControlledEffect(1, 0, Duration.EndOfTurn).setText("creatures you control get +1/+0 until end of turn"),
+                StaticFilters.FILTER_SPELL_A,
+                false
+        );
+
+        // Add scry effect to ability
+        ability.addEffect(new ScryEffect(1, false));
+
+        this.addAbility(ability);
+        this.addAbility(new GandalfWhiteRiderTriggeredAbility(new GandalfWhiteRiderDyingEffect(), true));
+    }
+
+    private GandalfWhiteRider(final GandalfWhiteRider card) {
+        super(card);
+    }
+
+    @Override
+    public GandalfWhiteRider copy() {
+        return new GandalfWhiteRider(this);
+    }
+}
+
+class GandalfWhiteRiderTriggeredAbility extends TriggeredAbilityImpl{
+
+    public GandalfWhiteRiderTriggeredAbility(Effect effect) {
+        this(effect, false);
+    }
+    public GandalfWhiteRiderTriggeredAbility(Effect effect, boolean optional){
+        super(Zone.ALL, effect, optional);
+        setTriggerPhrase("When {this} dies, ");
+    }
+
+    GandalfWhiteRiderTriggeredAbility(GandalfWhiteRiderTriggeredAbility ability){
+        super(ability);
+    }
+
+    @Override
+    public GandalfWhiteRiderTriggeredAbility copy(){
+        return new GandalfWhiteRiderTriggeredAbility(this);
+    }
+
+    @Override
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ZONE_CHANGE;
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        ZoneChangeEvent zEvent = (ZoneChangeEvent) event;
+        Permanent permanent = zEvent.getTarget();
+        if (permanent != null && zEvent.isDiesEvent()
+                && permanent.getId().equals(this.getSourceId())
+                &&
+                permanent.isOwnedBy(permanent.getControllerId())) {
+            return true;
+        }
+        return false;
+    }
+}
+
+class GandalfWhiteRiderDyingEffect extends OneShotEffect {
+
+    public GandalfWhiteRiderDyingEffect() {
+        super(Outcome.ReturnToHand);
+        staticText = "put it into its owner's library fifth from the top";
+    }
+
+    public GandalfWhiteRiderDyingEffect(final GandalfWhiteRiderDyingEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public GandalfWhiteRiderDyingEffect copy() {
+        return new GandalfWhiteRiderDyingEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+        Card card = (Card) source.getSourceObjectIfItStillExists(game);
```

Comment:
This doesn't look right... need to check the zcc (compare to the "permanentLeftBattlefield" value stored from DiesSourceTriggeredAbility)